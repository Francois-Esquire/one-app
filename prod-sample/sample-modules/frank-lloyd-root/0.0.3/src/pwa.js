/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

/* eslint-disable global-require */

export const defaultSplashScreen = 'http://localhost:3001/static/modules/frank-lloyd-root/0.0.3/assets/pwa-icon-512px.png';
export const defaultIcon = 'http://localhost:3001/static/modules/frank-lloyd-root/0.0.3/assets/pwa-icon-192px.png';
export const appleIcon = 'http://localhost:3001/static/modules/frank-lloyd-root/0.0.3/assets/pwa-icon-180px.png';

export const scope = '/';
// eslint-disable-next-line camelcase
export const start_url = '/success';
// eslint-disable-next-line camelcase
export const theme_color = '#FDB92D';

export default {
  enabled: true,
  scope,
  manifest: {
    scope,
    start_url,
    name: 'Frank Lloyd Root',
    short_name: 'frank lloyd root',
    display: 'standalone',
    theme_color,
    background_color: '#FFF',
    icons: [
      {
        src: appleIcon,
        type: 'image/png',
        sizes: '180x180',
      },
      {
        src: defaultIcon,
        type: 'image/png',
        sizes: '192x192',
      },
      {
        src: defaultSplashScreen,
        type: 'image/png',
        sizes: '512x512',
      },
    ],
  },
};
