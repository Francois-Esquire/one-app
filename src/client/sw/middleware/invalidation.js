/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// eslint-disable-next-line import/no-extraneous-dependencies
import {
  isCacheStorageSupported,
  isServiceWorker,
  getMetaData,
  setMetaData,
  remove,
} from '@americanexpress/one-service-worker';

import {
  getResourceTypeFromCacheName, getModuleInfoFromUrl, getLangPackInfoFromUrl, markForRemoval,
} from './utility';

export default function createInvalidationMiddleware() {
  if (!isCacheStorageSupported() || !isServiceWorker()) return function noop() {};

  return function invalidationMiddleware(event, context) {
    const { request, cacheName } = context.get();

    if (request && request instanceof Request) {
      const { url } = request;

      const type = getResourceTypeFromCacheName(cacheName);

      let info = {};

      switch (type) {
        case 'one-app':
          info = getLangPackInfoFromUrl(url);
          break;
        case 'module':
          info = getModuleInfoFromUrl(url);
          break;
        case 'language-pack':
          info = getLangPackInfoFromUrl(url);
          break;
        default:
          break;
      }

      const resourceKey = new Request(`/${type}/${info.name}`).url;
      const resourceMeta = {
        ...info,
        type,
        url,
      };

      event.waitUntil(
        getMetaData({
          url: resourceKey,
        }).then((meta) => {
          if (markForRemoval(meta, resourceMeta)) {
            event.waitUntil(remove(new Request(meta.url)));
          }

          return setMetaData({
            url: resourceKey,
            metadata: resourceMeta,
          });
        })
      );
    }
  };
}
