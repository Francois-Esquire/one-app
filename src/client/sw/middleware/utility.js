/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export function getConfig() {
  return JSON.parse(process.env.OSW_CONFIG || '{}');
}

export function getResourceTypeFromCacheName(cacheName = '') {
  const [, type = ''] = cacheName.match(/.*(one-app|language-pack|module)-cache$/) || [];
  return type;
}

export function getClientCacheRevisionFromUrl(url = '') {
  const [, key = ''] = url.match(/^https?.*(?:\.js\?clientCacheRevision=(.*))$/) || [];
  return key;
}

export function getLangPackInfoFromUrl(url = '') {
  // capture groups: name, version, locale, language, country, env,
  const info = url.match(/^https?.*\/(?<name>.*)\/(?<version>.*)\/(?<locale>(?<language>[a-z]{2,3})-(?<country>[A-Z]{1,4})?)\/(?<env>qa|integration|.*)\.json$/);

  if (info) {
    return info.groups;
  }

  return {};
}

export function getModuleInfoFromUrl(url = '') {
  // capture groups: checksum, name, version, bundle, legacy, revision
  const info = url.match(/^https?.*\/(?<checksum>.*)\/(?<name>.*)\/(?<version>.*)\/(?:.*)\2\.(?<bundle>(legacy\.)?browser)\.js(?:\?clientCacheRevision=(?<revision>.*))?$/);

  if (info) {
    return info.groups;
  }

  return {};
}

export function getAppInfoFromUrl(url = '') {
  // capture groups: version, name, i18n
  const info = url.match(/^https?.*\/_\/static\/app\/(?<version>.*)\/(?<name>(?<i18n>i18n)?.*)\.js$/);

  if (info) {
    return info.groups;
  }

  return {};
}

export function markForRemoval(cachedMetaRecord, newMetaRecord) {
  if (cachedMetaRecord.revision !== newMetaRecord.revision) return true;
  if (cachedMetaRecord.type === newMetaRecord.type) {
    if (cachedMetaRecord.name === newMetaRecord.name) {
      if (cachedMetaRecord.version !== newMetaRecord.version) return true;
      if (cachedMetaRecord.locale !== newMetaRecord.locale) return false;
    }
  }
  return false;
}
