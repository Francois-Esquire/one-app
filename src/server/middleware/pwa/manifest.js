/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

let webmanifest = null;
let webmanifestEnabled = false;

export function resetWebmanifest() {
  webmanifest = {
    name: 'One App',
    short_name: 'One App',
  };
  return webmanifest;
}

export function getWebmanifestEnabled() {
  return webmanifestEnabled;
}

export function getWebmanifest() {
  return webmanifest;
}

export function setWebmanifest(manifest) {
  if (manifest === null) webmanifest = null;
  else Object.assign(resetWebmanifest(), manifest);
  return getWebmanifest();
}

export function configureWebmanifest({ enabled = false, manifest } = {}) {
  webmanifestEnabled = enabled;
  if (enabled === false) setWebmanifest(null);
  else setWebmanifest(manifest);
}

export function webmanifestMiddleware() {
  return function webmanifestMiddlewareHandler(req, res, next) {
    if (!webmanifestEnabled) return next();
    return res
      .type('json')
      .send(getWebmanifest());
  };
}
