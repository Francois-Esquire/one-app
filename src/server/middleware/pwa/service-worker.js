/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import fs from 'fs';

let serviceWorkerScript = null;
let serviceWorkerScope = null;
let serviceWorkerType = null;
let serviceWorkerEnabled = false;

export function getServiceWorkerEnabled() {
  return serviceWorkerEnabled;
}

export function getServiceWorkerScope() {
  return serviceWorkerScope;
}

export function getServiceWorkerType() {
  return serviceWorkerType;
}

export function getServiceWorkerScript() {
  return serviceWorkerScript;
}

export function setServiceWorkerScript(value, scope) {
  if (scope) serviceWorkerScope = scope;
  else if (value) serviceWorkerScope = '/';
  else serviceWorkerScope = null;
  serviceWorkerScript = value;
  return serviceWorkerScript;
}

export function createServiceWorkerNoopScript() {
  // this file is created during build inside lib/server/middleware/pwa
  return fs.readFileSync([__dirname, 'scripts/sw.noop.js'].join('/')).toString();
}

export function createServiceWorkerScript() {
  // this file is created during build inside lib/server/middleware/pwa
  return fs.readFileSync([__dirname, 'scripts/sw.js'].join('/')).toString();
}

export function createServiceWorkerEscapeHatchScript() {
  return 'self.unregister();';
}

export function configureServiceWorker({
  enabled, escapeHatch, noop, scope,
} = {}) {
  if (escapeHatch) {
    serviceWorkerType = 'escape-hatch';
    serviceWorkerEnabled = true;
    setServiceWorkerScript(createServiceWorkerEscapeHatchScript(), scope);
  } else if (noop) {
    serviceWorkerType = 'noop';
    serviceWorkerEnabled = true;
    setServiceWorkerScript(createServiceWorkerNoopScript(), scope);
  } else if (enabled) {
    serviceWorkerType = 'standard';
    serviceWorkerEnabled = true;
    setServiceWorkerScript(createServiceWorkerScript(), scope);
  } else {
    serviceWorkerType = null;
    serviceWorkerEnabled = false;
    setServiceWorkerScript(null);
  }
}

export function serviceWorkerMiddleware() {
  return function serviceWorkerMiddlewareHandler(req, res, next) {
    if (serviceWorkerEnabled === false) return next();
    return res
      .type('js')
      .set('Service-Worker-Allowed', serviceWorkerScope)
      .set('Cache-Control', 'no-store, no-cache')
      .send(getServiceWorkerScript());
  };
}
