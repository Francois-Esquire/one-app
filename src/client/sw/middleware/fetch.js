/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// eslint-disable-next-line import/no-extraneous-dependencies
import { expiration, cacheRouter } from '@americanexpress/one-service-worker';

import createInvalidationMiddleware from './invalidation';
import { getConfig } from './utility';

export default function createFetchMiddleware(config = getConfig()) {
  return [
    {
      cacheName: 'one-app-cache',
      // eslint-disable-next-line no-useless-escape
      match: new RegExp(`^https?.*\/_\/static\/app\/${config.buildVersion}\/.*\\.js$`),
    },
    {
      cacheName: 'module-cache',
      // ends with .browser.js and matches the name
      match: /^https?.*\.(.*)\/(.*)\/(?:.*)\1(?:legacy\.)?browser\.js(?:\?clientCacheRevision=.*)?$/,
    },
    {
      cacheName: 'language-pack-cache',
      // any valid bcp 47 locale id
      match: /^https?.*\/.{2,3}-.{1,4}\/(qa|integration|.*)\.json$/,
    },
  ].map(cacheRouter).concat([
    createInvalidationMiddleware(),
    expiration(),
  ]);
}
