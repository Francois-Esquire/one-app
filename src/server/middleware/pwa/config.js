/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { validatePWAConfig } from './validation';
import { configureServiceWorker } from './service-worker';

const defaultPWAConfig = {
  serviceWorker: false,
  serviceWorkerRecoveryMode: false,
  serviceWorkerType: null,
  serviceWorkerScope: null,
};

let pwaConfig = { ...defaultPWAConfig };

function resetPWAConfig() {
  pwaConfig = { ...defaultPWAConfig };
}

function setPWAConfig(newConfiguration) {
  pwaConfig = newConfiguration;
  return pwaConfig;
}

export function getClientPWAConfig() {
  const { serviceWorker, serviceWorkerRecoveryMode, serviceWorkerScope } = pwaConfig;
  return {
    serviceWorker,
    serviceWorkerRecoveryMode,
    serviceWorkerScope,
    serviceWorkerScriptUrl: serviceWorker && '/_/pwa/service-worker.js',
  };
}

export function configurePWA(config) {
  // feature flag will not allow pwa/service-worker to be configured
  // it will default to a disabled state regardless if `appConfig.pwa` was provided
  if (process.env.ONE_SERVICE_WORKER !== 'true') {
    // eslint-disable-next-line no-param-reassign
    config = null;
    resetPWAConfig();
  }

  if (!config && pwaConfig.serviceWorker) {
    // if there was a previous configuration present, we want to gracefully
    // remove any remaining instances. We currently handle this client side
    // and would only need to reset the configuration when we want to decouple.
    resetPWAConfig();
  }

  const validatedConfig = config ? validatePWAConfig(config) : {};

  let enabled = false;
  let scope = null;
  let type = null;

  // the type dictates the service worker script to be used
  if (validatedConfig.escapeHatch) type = 'escape-hatch';
  else if (validatedConfig.recoveryMode) type = 'recovery';
  else if (validatedConfig.serviceWorker) type = 'standard';

  if (type) {
    // with a valid type, we can be sure that the service worker will be
    // ready to enable and assign a scope
    enabled = true;
    scope = validatedConfig.scope || '/';
  }

  configureServiceWorker({
    type,
    scope,
  });

  return setPWAConfig({
    serviceWorker: enabled,
    serviceWorkerScope: scope,
    serviceWorkerType: type,
    serviceWorkerRecoveryMode: ['escape-hatch', 'recovery'].includes(type),
  });
}
