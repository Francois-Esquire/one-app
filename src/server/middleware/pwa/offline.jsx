/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import { composeModules, RenderModule } from 'holocron';

import { getClientPWAConfig } from './config';
import {
  renderI18nScript, renderModuleScripts, serializeClientInitialState, modernBrowserChunkAssets,
} from '../sendHtml';
import renderModuleStyles from '../../utils/renderModuleStyles';
import { renderShell, renderInitialState, renderScriptTag } from '../../utils/renderShell';
import { getClientStateConfig } from '../../utils/stateConfig';
import { renderForStaticMarkup } from '../../utils/reactRendering';
import jsonStringifyForScript from '../../utils/jsonStringifyForScript';
import readJsonFile from '../../utils/readJsonFile';

const { buildVersion } = readJsonFile('../../../.build-meta.json');
const bundleIntegrity = readJsonFile('../../../bundle.integrity.manifest.json');

export default function createOfflineShellMiddleware() {
  const bundle = 'browser';
  const isDevelopment = process.env.NODE_ENV === 'development';

  return async function offlineShellMiddleware(req, res) {
    const { store, clientModuleMapCache } = req;
    const { scriptNonce } = res;
    const clientConfig = getClientStateConfig();
    const pwaMetaData = getClientPWAConfig();

    const cdnUrl = `${clientConfig.cdnUrl || '/_/static/'}app/${buildVersion}`;
    const rootModuleName = store.getState().getIn(['config', 'rootModuleName']);
    const activeLocale = store.getState().getIn(['intl', 'activeLocale']);

    const [app, appVendors, runtime, , vendors] = ['app.js'].concat(modernBrowserChunkAssets).map((resource) => ({
      src: `${cdnUrl}/${resource}`,
      integrity: bundleIntegrity[resource],
    })).map(({ src, integrity }) => renderScriptTag(src, [`integrity="${integrity}"`]));

    await store.dispatch(composeModules([{ name: rootModuleName }]));

    const { renderedString: appHtml, helmetInfo } = renderForStaticMarkup(
      <Provider store={store}>
        <RenderModule moduleName={rootModuleName} />
      </Provider>
    );

    const {
      htmlAttributes, bodyAttributes, base, title, meta, link, style, script, noscript,
    } = helmetInfo
      ? Object.entries(helmetInfo)
        .reduce((object, [key, value]) => ({
          ...object,
          [key]: value && typeof value === 'object' ? value.toString() : value || '',
        }), {})
      : {};

    const clientInitialState = store.getState();

    const styles = [].concat(renderModuleStyles(store), style).join('');
    const scripts = [].concat([
      appVendors,
      runtime,
      vendors,
      renderI18nScript(clientInitialState, cdnUrl),
      renderModuleScripts({
        clientInitialState,
        moduleMap: clientModuleMapCache[bundle],
        isDevelopmentEnv: isDevelopment,
        bundle,
      }),
      app,
    ], script).join('');
    const initialState = renderInitialState({
      config: {
        isDevelopment,
        scriptNonce,
      },
      // dictates if the client app should start rendering with
      // either ReactDOM method listed below
      __render_mode__: '\'render\'',
      // we expect the cdnUrl = 'https://example.com/cdn' || '/_/static/'
      // and we combine with the buildPath and add a trailing slash
      __webpack_public_path__: `'${cdnUrl}/'`,
      // we toggle between `browser` and `legacyBrowser` depending on `bundleType`
      __holocron_module_bundle_type__: '\'browser\'',
      // The client module map is derived from the original module map
      // and is expected to be stringified here, we do not wrap it in quotes
      // as we want it directly usable
      __CLIENT_HOLOCRON_MODULE_MAP__: jsonStringifyForScript(
        clientModuleMapCache.browser
      ),
      // The client initial state after it has serialized
      // the immutable state and stringified
      __INITIAL_STATE__: jsonStringifyForScript(
        serializeClientInitialState(clientInitialState)),
      // used to configure the client-side PWA initialization
      __pwa_metadata__: jsonStringifyForScript(pwaMetaData),
    }, {
      manifest: pwaMetaData.webManifestUrl ? [pwaMetaData.webManifestUrl] : null,
    });

    res
      .status(200)
      .type('html')
      .send(
        renderShell({
          lang: activeLocale,
          appHtml,
          scripts,
          styles,
          meta,
          base,
          title,
          links: link,
          htmlAttrs: htmlAttributes,
          bodyAttrs: bodyAttributes,
          noScripts: noscript,
          initialState,
        })
      );
  };
}
