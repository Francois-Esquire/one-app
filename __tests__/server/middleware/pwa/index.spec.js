/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */


import request from 'supertest';
import express from 'express';

import {
  createServiceWorkerEscapeHatchScript,
  readServiceWorkerRecoveryScript,
  readServiceWorkerScript,
  serviceWorkerMiddleware,
} from '../../../../src/server/middleware/pwa/service-worker';
import * as pwa from '../../../../src/server/middleware/pwa';

jest.mock('fs', () => ({
  readFileSync: (filePath) => (filePath.endsWith('noop.js') ? '[service-worker-noop-script]' : '[service-worker-script]'),
}));

describe('PWA', () => {
  test('should export components and router', () => {
    expect(Object.entries(pwa)).toMatchSnapshot();
  });
});

describe('PWA routing', () => {
  const defaultMatchAll = jest.fn((req, res) => {
    res.sendStatus(404);
  });

  const makeGetFrom = (app) => (url) => new Promise((resolve, reject) => {
    request(app)
      .get(url)
      .end((err, res) => {
        if (err) reject(err);
        else resolve(res);
      });
  });

  const workerPath = '/_/pwa/service-worker.js';
  const mockWorker = readServiceWorkerScript();
  const mockRecoveryWorker = readServiceWorkerRecoveryScript();
  const mockEscapeHatchWorker = createServiceWorkerEscapeHatchScript().toString();
  const mockServiceWorkerScope = '/nested/scope';

  beforeEach(() => jest.clearAllMocks());
  beforeAll(() => {
    process.env.ONE_SERVICE_WORKER = true;
    pwa.configurePWA({
      serviceWorker: true,
    });
  });

  const app = express()
    .get(workerPath, serviceWorkerMiddleware())
    .get('*', defaultMatchAll);
  const get = makeGetFrom(app);

  [
    [workerPath, 'application/javascript; charset=utf-8', mockWorker],
  ].forEach((pair) => {
    const [route, contentType, mockResponse] = pair;

    it(`should respond with a status of 200 at ${route}`, async () => {
      expect.assertions(3);
      const response = await get(route);
      expect(response.status).toBe(200);
      expect(response.headers).toHaveProperty('content-type', contentType);
      expect(response.text).toBe(mockResponse);
    });
  });

  describe('disabled', () => {
    beforeAll(() => {
      pwa.configurePWA({
        serviceWorker: false,
      });
    });

    [
      [workerPath, 404],
    ].forEach((pair) => {
      const [route, status] = pair;

      it(`should respond with a status of ${status} when disabled at ${route}`, async () => {
        expect.assertions(1);
        const response = await get(route);
        expect(response.status).toBe(status);
      });
    });
  });

  describe('recovery mode', () => {
    beforeAll(() => {
      pwa.configurePWA({
        recoveryMode: true,
      });
    });

    [
      [workerPath, 'application/javascript; charset=utf-8', mockRecoveryWorker],
    ].forEach((pair) => {
      const [route, contentType, mockResponse] = pair;

      it(`should respond with recovery asset at ${route}`, async () => {
        expect.assertions(3);
        const response = await get(route);
        expect(response.status).toBe(200);
        expect(response.headers).toHaveProperty('content-type', contentType);
        expect(response.text).toBe(mockResponse);
      });
    });
  });

  describe('escape hatch worker', () => {
    beforeAll(() => {
      pwa.configurePWA({
        escapeHatch: true,
      });
    });

    [
      [workerPath, 'application/javascript; charset=utf-8', mockEscapeHatchWorker],
    ].forEach((pair) => {
      const [route, contentType, mockResponse] = pair;

      it(`should respond with a status of 404 when disabled at ${route}`, async () => {
        expect.assertions(3);
        const response = await get(route);
        expect(response.status).toBe(200);
        expect(response.headers).toHaveProperty('content-type', contentType);
        expect(response.text).toBe(mockResponse);
      });
    });
  });

  describe('setting scope', () => {
    beforeAll(() => {
      pwa.configurePWA({
        serviceWorker: true,
        scope: mockServiceWorkerScope,
      });
    });

    [
      [workerPath, mockServiceWorkerScope, mockWorker],
    ].forEach((pair) => {
      const [route, mockedScope, mockResponse] = pair;

      it(`should respond with a status of 404 when disabled at ${route}`, async () => {
        expect.assertions(3);
        const response = await get(route);
        expect(response.status).toBe(200);
        expect(response.headers).toHaveProperty('service-worker-allowed', mockedScope);
        expect(response.text).toBe(mockResponse);
      });
    });
  });
});
