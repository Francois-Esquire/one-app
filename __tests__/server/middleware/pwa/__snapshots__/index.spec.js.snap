// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PWA middleware should export components and middleware 1`] = `
Array [
  Array [
    "configurePWA",
    [Function],
  ],
  Array [
    "getClientPWAConfig",
    [Function],
  ],
  Array [
    "getPWAConfig",
    [Function],
  ],
  Array [
    "webmanifestMiddleware",
    [Function],
  ],
  Array [
    "serviceWorkerMiddleware",
    [Function],
  ],
]
`;
